File: Adramelech/Commands/Slash/Lookup.cs
using System.Diagnostics.CodeAnalysis;
using System.Text.RegularExpressions;
using Adramelech.Common;
using Adramelech.Configuration;
using Adramelech.Extensions;
using Adramelech.Utilities;
using Discord;
using Discord.Interactions;
using Discord.WebSocket;
using Flurl;

namespace Adramelech.Commands.Slash;

public class Lookup(Config config, HttpUtils httpUtils)
    : InteractionModuleBase<SocketInteractionContext<SocketSlashCommand>>
{
    [SlashCommand("lookup", "Lookup a ip or domain")]
    public async Task LookupAsync([Summary("target", "The target to lookup")] string target)
    {
        await DeferAsync();

        var ip = LookupHelper.IpRegex().Match(target).Success
            ? target
            : await GetIpFromDomain(target);
        if (ip.IsFailure)
        {
            await Context.SendError("Failed to get ip from domain", true);
            return;
        }

        var response = await httpUtils.GetAsync<Whois>($"https://ipwho.is/{ip.Value}", "curl");
        if (response.IsDefault() || !response.Success)
        {
            await Context.SendError("Failed to get whois information", true);
            return;
        }

        var mainField = $"""
                         **IP:** {ip.Value}
                         **Domain:** {(target != ip.Value ? target : "N/A")}
                         **Type:** {response.Type}
                         """;
        var locationField = $"""
                             **Continent:** {response.Continent}
                             **Country:** {response.Country} :flag_{response.CountryCode.ToLower()}:
                             **Region:** {response.Region}
                             **City:** {response.City}
                             **Latitude:** {response.Latitude}
                             **Longitude:** {response.Longitude}
                             **Postal:** {response.Postal}
                             """;
        var connectionField = $"""
                               **ASN:** {response.Connection.Asn}
                               **Organization:** {response.Connection.Org}
                               **ISP:** {response.Connection.Isp}
                               **Domain:** {response.Connection.Domain}
                               """;
        var timezoneField = $"""
                             **ID:** {response.Timezone.Id}
                             **Offset:** {response.Timezone.Offset}
                             **UTC:** {response.Timezone.Utc}
                             """;

        var mapsUrl = new Url("https://www.google.com")
            .AppendPathSegments("maps", "search", "/")
            .SetQueryParam("api", 1)
            .SetQueryParam("query", $"{response.Latitude},{response.Longitude}");

        await FollowupAsync(
            embed: new EmbedBuilder()
                .WithColor(config.EmbedColor)
                .WithTitle("Lookup")
                .WithDescription("For best results, search by ip")
                .AddField(":zap: Main", mainField)
                .AddField(":earth_americas: Location", locationField)
                .AddField(":satellite: Connection", connectionField)
                .AddField(":clock1: Timezone", timezoneField)
                .WithFooter("Powered by ipwho.is and da.gd")
                .Build(),
            components: new ComponentBuilder()
                .WithButton("Open in Google Maps", style: ButtonStyle.Link, url: mapsUrl.ToString(),
                    emote: new Emoji("🌎"))
                .Build());
    }

    private async Task<Result<string>> GetIpFromDomain(string domain)
    {
        var response = await httpUtils.GetAsync<string>($"https://da.gd/host/{domain}");
        if (response.IsNullOrEmpty()) return new Exception("Failed to get ip from domain");

        response = response!.Trim();

        if (response.StartsWith("No"))
            return new Exception("No ip found for domain");

        return response.Contains(',') ? response[..response.IndexOf(',')] : response;
    }

    [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Local")]
    private struct Whois
    {
        public bool Success { get; set; }
        public string Type { get; set; }
        public string Continent { get; set; }
        public string Country { get; set; }
        public string CountryCode { get; set; }
        public string Region { get; set; }
        public string City { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public string Postal { get; set; }
        public SConnection Connection { get; set; }
        public STimezone Timezone { get; set; }

        internal struct SConnection
        {
            public int Asn { get; set; }
            public string Org { get; set; }
            public string Isp { get; set; }
            public string Domain { get; set; }
        }

        internal struct STimezone
        {
            public string Id { get; set; }
            public int Offset { get; set; }
            public string Utc { get; set; }
        }
    }
}

// ReSharper disable once ClassNeverInstantiated.Global
internal partial class LookupHelper
{
    [GeneratedRegex(@"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b")]
    public static partial Regex IpRegex();
}

File: bot/src/commands/slash/lookup.ts
import {
  ButtonStyle,
  ComponentType,
  MessageFlags,
  SlashCommandBuilder,
  type ChatInputCommandInteraction,
} from 'discord.js';
import ky from 'ky';
import z from 'zod';
import type { Command } from '~/types/command';
import { sendError } from '~/utils/sendError';
import config from '~/config';
import { stripIndents } from 'common-tags';
import { errAsync, fromAsyncThrowable, ok, okAsync } from 'neverthrow';

const schema = z.object({
  success: z.boolean(),
  type: z.union([z.literal('IPv4'), z.literal('IPv6')]),
  continent: z.string(),
  country: z.string(),
  country_code: z.string().length(2),
  region: z.string(),
  city: z.string(),
  latitude: z.number(),
  longitude: z.number(),
  postal: z.string(),
  connection: z.object({
    asn: z.number(),
    org: z.string(),
    isp: z.string(),
    domain: z.string(),
  }),
  timezone: z.object({
    id: z.string(),
    offset: z.number(),
    utc: z.string(),
  }),
});

export const command = <Command>{
  data: new SlashCommandBuilder()
    .setName('lookup')
    .setDescription('Lookup a domain or IP address')
    .addStringOption((option) =>
      option
        .setName('target')
        .setDescription('The domain or IP address to lookup')
        .setRequired(true),
    ),
  cooldown: true,
  uses: ['ipwhois.io', 'da.gd'],
  async execute(intr: ChatInputCommandInteraction) {
    await intr.deferReply();

    const target = intr.options.getString('target', true);

    const ip = z.union([z.ipv4(), z.ipv6()]).safeParse(target).success
      ? ok(target)
      : await getIpFromDomain(target);
    if (ip.isErr()) return await sendError(intr, ip.error);

    const response = await fromAsyncThrowable(
      ky.get(`https://ipwho.is/${ip.value}`, {
        headers: {
          'User-Agent': 'curl', // shh... you didn't see this :3
        },
      }).json,
      (error) =>
        error instanceof Error
          ? error.message
          : 'Failed to fetch IP information',
    )().andThen((json) => {
      const parsed = schema.safeParse(json);
      return parsed.success
        ? okAsync(parsed.data)
        : errAsync('Failed to parse response');
    });
    if (response.isErr()) return await sendError(intr, response.error);
    const data = response.value;

    const mapsUrl = new URL('https://www.google.com/maps/search/');
    mapsUrl.searchParams.set('api', '1');
    mapsUrl.searchParams.set('query', `${data.latitude},${data.longitude}`);

    await intr.followUp({
      flags: MessageFlags.IsComponentsV2,
      components: [
        {
          type: ComponentType.Container,
          accent_color: config.EMBED_COLOR,
          components: [
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              # Lookup
              **IP:** ${ip.value}
              **Domain:** ${target === ip.value ? 'None' : target}
              **Type:** ${data.type}
              `,
            },
            { type: ComponentType.Separator },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ## Location
              **Continent:** ${data.continent}
              **Country:** ${data.country} :flag_${data.country_code.toLowerCase()}:
              **Region:** ${data.region}
              **City:** ${data.city}
              **Latitude:** ${data.latitude}
              **Longitude:** ${data.longitude}
              **Postal Code:** ${data.postal}
              `,
            },
            {
              type: ComponentType.ActionRow,
              components: [
                {
                  type: ComponentType.Button,
                  style: ButtonStyle.Link,
                  label: 'Open in Google Maps',
                  url: mapsUrl.toString(),
                  emoji: { name: '🌎' },
                },
              ],
            },
            { type: ComponentType.Separator },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ## Connection
              **ASN:** ${data.connection.asn}
              **Organization:** ${data.connection.org}
              **ISP:** ${data.connection.isp}
              **Domain:** ${data.connection.domain}
              `,
            },
            { type: ComponentType.Separator },
            {
              type: ComponentType.TextDisplay,
              content: stripIndents`
              ## Timezone
              **ID:** ${data.timezone.id}
              **Offset:** ${data.timezone.offset}
              **UTC:** ${data.timezone.utc}
              `,
            },
            {
              type: ComponentType.TextDisplay,
              content: '> Powered by ipwhois.io and da.gd',
            },
          ],
        },
      ],
    });
  },
};

const getIpFromDomain = (domain: string) =>
  fromAsyncThrowable(ky.get(`https://da.gd/host/${domain}`).text, (error) =>
    error instanceof Error ? error.message : 'Failed to get IP from domain',
  )().andThen((text) => {
    const res = text.trim();
    if (!res || res.startsWith('No'))
      return errAsync('Failed to get IP from domain');
    return okAsync(
      res.includes(',') ? res.substring(0, res.indexOf(',')) : res,
    );
  });
